clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractHeartbeatRequestManager.java

```diff
@@ -195,6 +197,32 @@ public NetworkClientDelegate.PollResult poll(long currentTimeMs) {
      */
     public abstract AbstractMembershipManager<R> membershipManager();
 
+    /**
+     * Generate a heartbeat request to leave the group if the state is still LEAVING when this is
+     * called to close the consumer.
```

Comment:
Is it possible that there's already a 'leave group' request that was sent? Won't `AsyncKafkaConsumer.releaseAssignmentAndLeaveGroup()` be called before before the request managers are closed.

Or does it not matter if we send more than one leave group request?

Reply:
There couldn't be a previous leave request, given that the state is LEAVING/PREPARE_LEAVING (basically the state machine controls that, when leaving, it transitions to UNSUBSCRIBED as soon as the request is sent, [here](https://github.com/apache/kafka/blob/62dc982ce9ce76b299a628d50648e3529fd88e63/clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractMembershipManager.java#L646))

Reply:
👍 Thanks for the explanation 😄